File: backend/src/routes/billing.ts
import { Hono } from "hono";
import { z } from "zod";
import type { AuthVariables } from "../lib/auth.js";
import { mapCreditTransaction } from "../lib/mappers.js";
import type { Env } from "../lib/env.js";
import { createAdminClient } from "../lib/supabase.js";
import { createBillingService, BillingNotConfiguredError } from "../services/billing.js";

const checkoutSchema = z.object({
  packId: z.string().min(1),
  successUrl: z.string().url(),
  cancelUrl: z.string().url(),
});

export function billingRoutes(env: Env) {
  const admin = createAdminClient(env);
  const { billing } = createBillingService(admin);

  return new Hono<{ Variables: AuthVariables }>()
    .get("/transactions", async (c) => {
      const supabase = c.get("supabase");
      const userId = c.get("userId");

      const { data, error } = await supabase
        .from("credit_transactions")
        .select("*")
        .eq("user_id", userId)
        .order("created_at", { ascending: false });

      if (error) return c.json({ error: error.message }, 400);

      return c.json({
        transactions: (data ?? []).map(mapCreditTransaction),
      });
    })
    .get("/packs", async (c) => {
      const packs = await billing.listPacks();
      return c.json({ packs });
    })
    .post("/checkout", async (c) => {
      const userId = c.get("userId");
      const body = checkoutSchema.parse(await c.req.json());

      try {
        const session = await billing.createCheckoutSession(userId, body);
        return c.json(session);
      } catch (error) {
        if (error instanceof BillingNotConfiguredError) {
          return c.json({ error: error.message }, 501);
        }
        throw error;
      }
    });
}

export function billingWebhookRoutes(env: Env) {
  const admin = createAdminClient(env);
  const { billing } = createBillingService(admin);

  return new Hono().post("/webhook", async (c) => {
    const signature = c.req.header("stripe-signature");
    const payload = Buffer.from(await c.req.arrayBuffer());

    try {
      await billing.handleStripeWebhook(payload, signature);
      return c.json({ received: true });
    } catch (error) {
      if (error instanceof BillingNotConfiguredError) {
        return c.json({ error: error.message }, 501);
      }
      throw error;
    }
  });
}

File: backend/src/captionovo/routes/billing.py
from fastapi import APIRouter, Depends, HTTPException, Request

from captionovo.auth import AuthContext, get_auth_context
from captionovo.config import Settings, get_settings
from captionovo.mappers import map_credit_transaction
from captionovo.schemas import CheckoutBody
from captionovo.services.billing import BillingNotConfiguredError, create_billing_service
from captionovo.supabase_client import create_admin_client

router = APIRouter(prefix="/billing", tags=["billing"])
webhook_router = APIRouter(prefix="/billing", tags=["billing"])


def _pack_to_response(pack) -> dict:
    result = {
        "id": pack.id,
        "name": pack.name,
        "credits": pack.credits,
        "priceCents": pack.price_cents,
    }
    if pack.stripe_price_id:
        result["stripePriceId"] = pack.stripe_price_id
    return result


@router.get("/transactions")
async def list_transactions(auth: AuthContext = Depends(get_auth_context)):
    result = (
        auth.supabase.table("credit_transactions")
        .select("*")
        .eq("user_id", auth.user_id)
        .order("created_at", desc=True)
        .execute()
    )
    if result.data is None and hasattr(result, "error") and result.error:
        raise HTTPException(status_code=400, detail={"error": str(result.error)})

    return {"transactions": [map_credit_transaction(row) for row in (result.data or [])]}


@router.get("/packs")
async def list_packs(settings: Settings = Depends(get_settings)):
    admin = create_admin_client(settings)
    billing = create_billing_service(admin)["billing"]
    packs = billing.list_packs()
    return {"packs": [_pack_to_response(p) for p in packs]}


@router.post("/checkout")
async def checkout(body: CheckoutBody, auth: AuthContext = Depends(get_auth_context)):
    admin = create_admin_client(get_settings())
    billing = create_billing_service(admin)["billing"]

    from captionovo.domain.billing import CheckoutSessionRequest

    request = CheckoutSessionRequest(
        pack_id=body.pack_id,
        success_url=str(body.success_url),
        cancel_url=str(body.cancel_url),
    )

    try:
        session = billing.create_checkout_session(auth.user_id, request)
        return session
    except BillingNotConfiguredError as exc:
        raise HTTPException(status_code=501, detail={"error": str(exc)}) from exc


@webhook_router.post("/webhook")
async def stripe_webhook(request: Request, settings: Settings = Depends(get_settings)):
    signature = request.headers.get("stripe-signature")
    payload = await request.body()

    admin = create_admin_client(settings)
    billing = create_billing_service(admin)["billing"]

    try:
        billing.handle_stripe_webhook(payload, signature)
        return {"received": True}
    except BillingNotConfiguredError as exc:
        raise HTTPException(status_code=501, detail={"error": str(exc)}) from exc
